semgrep-core/matching/Rules_filter.ml

```diff
@@ -0,0 +1,155 @@
+(* Yoann Padioleau
+ *
+ * Copyright (C) 2020 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file license.txt.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * license.txt for more details.
+ *)
+open Common
+module Flag = Flag_semgrep
+module R = Rule
+module V = Visitor_AST
+open AST_generic
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* Rules filtering using regexps.
+ *
+ * It is useless to run the engine on rules containing specific IDs
+ * (e.g., eval) that are never mentioned in the file.
+ *
+ * We did something similar in Coccinelle I think. This also has been
+ * mentioned many times (by Clint, HN, etc.).
+ *
+ * notes: I tried to use the ocaml-re (Re) regexp libray instead of Str
+ * because I thought it would be faster, and because it offers regexp
+ * combinators (alt, rep, etc.) which might be useful at some point to
+ * handle patterns containing explicit DisjExpr. However when running
+ * on Zulip codebase with zulip semgrep rules, Str is actually faster
```

Comment:
😛

Reply:
I haven't used Re. Based on the [project description](https://github.com/ocaml/ocaml-re), it's faster than Pcre and Str only on large regexps. This is presumably due to using an algorithm with polynomial rather than exponential asymptotic complexity, but with a greater startup cost due to the construction of an automaton.

Reply:
I was reading this article on the subject the other day: https://swtch.com/~rsc/regexp/regexp1.html
It's interesting to grasp how the path-merging algorithm works and why it has not problem handling regular expressions made of many nodes.